added env TARGET="HASWELL" to openblas build

zlscherr opened the pull request:
Openblas currently builds with DYNAMIC_ARCH=1, but
if you don't specify the target CPU then the bottled
version can only be run on any machine whose CPU is
as least as new as the CPU which built openblas.

By specifying TARGET="HASWELL" older machine should
be able to use the bottled version of openblas
without having to build from scratch.

See https://github.com/Homebrew/discussions/discussions/54
for a discussion about how this error can cause
libraries like numpy, which are built on openblas, to
segfault.

- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

jonchang commented:
What targets are available? I'm worried Haswell is too new for many of our users. When bottling, we use Nehalem on newer macOS and core2 for older ones. 

https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/mac/hardware.rb

zlscherr commented:
> What targets are available? I'm worried Haswell is too new for many of our users. When bottling, we use Nehalem on newer macOS and core2 for older ones.
> 
> https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/mac/hardware.rb

I might need to dig more into my issues with openBLAS as discussed at [Homebrew/discussions#54](https://github.com/Homebrew/discussions/discussions/54).

When I added TARGET=HASWELL then my numpy worked correctly.  However, if I tried to build with
TARGET=NEHALEM then I end up having the same issue.  Maybe I need to ask people who know
more about openBLAS what exactly is going on.  I was under the impression that building with an older
TARGET and DYNAMIC_ARCH=1 would solve my problems, but it doesn't seem to be working the way
I had thought it should.

SMillerDev commented on Formula/openblas.rb:
```diff
@@ -23,6 +23,7 @@ def install
     ENV["DYNAMIC_ARCH"] = "1"
     ENV["USE_OPENMP"] = "1"
     ENV["NO_AVX512"] = "1"
+    ENV["TARGET"] = "HASWELL"
```
This would do what homebrew uses for the bottles

```suggestion
    ENV["TARGET"] = Hardware.oldest_cpu
```

zlscherr replied:
I talked to the openBLAS people and my specific issues are apparently caused by a bug in 3.10 which will be fixed in 3.11.

But it's still important to set the TARGET because I think people can still have issues if the machine bottling the libraries has a newer processor than what they use.  So Hardware.oldest_cpu seems wise.

isuruf replied:
You need to make sure `Hardware.oldest_cpu` is something OpenBLAS understands though.

jonchang commented:
Ok, I've caught up on the issues over at the openblas repo. Let's do the following:

Add this patch in a `patch do` block https://github.com/xianyi/OpenBLAS/pull/2729
For the TARGET variable, do something like this:

```ruby
ENV["TARGET"] = case Hardware.oldest_cpu
when :arm_vortex_tempest
  "VORTEX"
else
  Hardware.oldest_cpu.upcase
end
```

zlscherr commented:
I added fixed up the TARGET and added the patch Jonathan mentioned.

jonchang commented:
> I added fixed up the TARGET and added the patch Jonathan mentioned.

Can you confirm that these two changes in concert fix the numpy issues? (Ideally, by installing the bottle from the workflow artifact)

zlscherr commented:
> > I added fixed up the TARGET and added the patch Jonathan mentioned.
> 
> Can you confirm that these two changes in concert fix the numpy issues? (Ideally, by installing the bottle from the workflow artifact)

Downloaded and tested the bottle and it is working correctly on my Catalina machine.  I don't know enough about the bottling process, but I assume the bottle do block and code signing happens automatically after the change is merged? When I tested the bottled artifact I needed to manually code sign it after installing, but numpy worked without crashing!

SMillerDev commented:
If it works fine for you I believe it'll work for everyone. Yes, the bottles will be done upon merging by the bot.

SMillerDev reviewed (approved):
Thanks @zlscherr  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉
